src/freenas/usr/local/bin/snmp-agent.py

```diff
@@ -527,6 +550,126 @@ def run(self):
+     "out_last_10s", "in_last_40s", "out_last_40s"]
+)
+
+
+class IftopInterfaceThread(threading.Thread):
+    def __init__(self, interface, stats):
+        super().__init__()
+
+        self.daemon = True
+
+        self.interface = interface
+        self.stats = stats
+
+    def run(self):
+        while True:
+            try:
+                master, slave = pty.openpty()
+                try:
+                    p = subprocess.Popen(
+                        [
+                            "iftop",
+                            "-n",           # Don't do hostname lookups
+                            "-N",           # Do not resolve port number to service names
+                            "-P",           # Turn on port display
+                            "-B",           # Display bandwidth rates in bytes/sec rather than bits/sec
+                            "-i", self.interface,
+                            "-t",           # use text interface without ncurses
+                            "-L", "10",     # number of lines to print
+                        ],
+                        stdin=subprocess.DEVNULL,
+                        stdout=slave,
+                        stderr=subprocess.DEVNULL,
+                        encoding="utf-8",
+                        errors="ignore",
+                        preexec_fn=os.setsid,
+                    )
+                    try:
+                        os.close(slave)
+                        with os.fdopen(master, "r", encoding="utf-8", errors="ignore") as f:
+                            data = []
+                            while True:
+                                line = f.readline()
+                                if not line:
+                                    break
+
+                                if line.startswith("-" * 80):
+                                    if (
+                                            len(data) % 2 == 0 and
+                                            all(len(v.split()) == (7 if i % 2 == 0 else 6) for i, v in enumerate(data))
+                                    ):
+                                        self.stats[self.interface] = self._handle_data(data)
+
+                                    data = []
+                                else:
+                                    data.append(line)
+                    finally:
+                        p.wait()
+                        print(f"iftop for {self.interface} returned {p.returncode}")
+                finally:
+                    with contextlib.suppress(OSError):
+                        os.close(master)
+                    with contextlib.suppress(OSError):
+                        os.close(slave)
+            except Exception as e:
+                print(f"Failed to run iftop for {self.interface}: {e!r}")
+
+            time.sleep(10)
+
+    def _handle_data(self, data):
+        stats = []
+        for line1, line2 in zip(data[::2], data[1::2]):
```

Comment:
Interesting functionality, I wasnt aware of `::` :P

Reply:
I think I use this for the first time in my life.